ui/ozone/platform/wayland/wayland_connection.h

```diff
@@ -196,6 +204,7 @@ class WaylandConnection : public PlatformEventSource,
   std::unique_ptr<WaylandDataDeviceManager> data_device_manager_;
   std::unique_ptr<WaylandDataDevice> data_device_;
   std::unique_ptr<WaylandDataSource> data_source_;
```

Comment:
is it really necessary to have two WaylandDataSources?

Reply:
I think CopyAndPaste and Dnd could work at the same time. For instance, a user copies something, clipboard holds it and he could try DnD. When 'WaylandDataSource::OnCancel' or 'WaylandDataSource::OnDnDFinished' can not destroy the instance because clipboard holds copied data yet. I also checked Weston code and it manages each data source for DnD, editor and terminal.